File: 9-binary-search-tree/code.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Struktur Node untuk Binary Search Tree
typedef struct BSTNode {
    int data;
    struct BSTNode* left;
    struct BSTNode* right;
} BSTNode;

// Fungsi untuk membuat node baru
BSTNode* createNode(int data) {
    BSTNode* newNode = (BSTNode*)malloc(sizeof(BSTNode));
    if (newNode == NULL) {
        printf("Memory allocation failed\n");
        exit(1);
    }
    newNode->data = data;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
}

// Fungsi untuk menyisipkan node ke dalam BST
BSTNode* insert(BSTNode* root, int data) {
    if (root == NULL) {
        return createNode(data);
    }
    
    if (data < root->data) {
        root->left = insert(root->left, data);
    } else if (data > root->data) {
        root->right = insert(root->right, data);
    }
    
    return root;
}

// Fungsi untuk mencari nilai dalam BST
BSTNode* search(BSTNode* root, int key) {
    if (root == NULL || root->data == key) {
        return root;
    }
    
    if (key < root->data) {
        return search(root->left, key);
    }
    
    return search(root->right, key);
}

// Fungsi untuk membersihkan BST
void freeBST(BSTNode* root) {
    if (root != NULL) {
        freeBST(root->left);
        freeBST(root->right);
        free(root);
    }
}

int main() {
    clock_t start, end;
    double cpu_time_used;
    
    // Inisialisasi BST
    BSTNode* root = NULL;
    
    start = clock();
    
    // Operasi pada BST
    // Menambahkan node (misal 1000 node)
    for (int i = 0; i < 1000; i++) {
        int value = rand() % 10000; // Nilai acak untuk menghindari BST yang tidak seimbang
        root = insert(root, value);
    }
    
    // Mencari beberapa nilai
    for (int i = 0; i < 100; i++) {
        int key = rand() % 10000;
        BSTNode* result = search(root, key);
        if (result != NULL) {
            printf("Found %d\n", key);
        } else {
            printf("%d not found\n", key);
        }
    }
    
    freeBST(root);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Waktu eksekusi: %f detik\n", cpu_time_used);
    
    return 0;
}

File: 9-binary-search-tree/code.rs
use std::time::Instant;

// Struktur Node untuk BST
struct Node {
    data: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

// Implementasi untuk Node
impl Node {
    // Membuat node baru
    fn new(data: i32) -> Self {
        Node {
            data,
            left: None,
            right: None,
        }
    }
}

// Struktur untuk mengelola BST
struct BST {
    root: Option<Box<Node>>,
}

// Implementasi untuk BST
impl BST {
    // Membuat BST baru
    fn new() -> Self {
        BST {
            root: None,
        }
    }
    
    // Menyisipkan node ke dalam BST
    fn insert(&mut self, data: i32) {
        self.root = Self::insert_recursive(self.root.take(), data);
    }
    
    // Helper function untuk insert
    fn insert_recursive(node: Option<Box<Node>>, data: i32) -> Option<Box<Node>> {
        match node {
            None => Some(Box::new(Node::new(data))),
            Some(mut node) => {
                if data < node.data {
                    node.left = Self::insert_recursive(node.left.take(), data);
                } else if data > node.data {
                    node.right = Self::insert_recursive(node.right.take(), data);
                }
                Some(node)
            }
        }
    }
    
    // Mencari nilai dalam BST
    fn search(&self, key: i32) -> bool {
        Self::search_recursive(&self.root, key)
    }
    
    // Helper function untuk search
    fn search_recursive(node: &Option<Box<Node>>, key: i32) -> bool {
        match node {
            None => false,
            Some(node) => {
                if node.data == key {
                    true
                } else if key < node.data {
                    Self::search_recursive(&node.left, key)
                } else {
                    Self::search_recursive(&node.right, key)
                }
            }
        }
    }
}

fn main() {
    let start = Instant::now();
    
    // Inisialisasi BST
    let mut bst = BST::new();
    
    // Menambahkan node (misal 1000 node dengan nilai deterministik)
    for i in 0..1000 {
        let value = (i * 83) % 10000;
        bst.insert(value);
    }
    
    // Mencari beberapa nilai
    for i in 0..100 {
        let key = (i * 97) % 10000;
        let found = bst.search(key);
        if found {
            println!("Found {}", key);
        } else {
            println!("{} not found", key);
        }
    }
    
    // BST akan dibersihkan secara otomatis ketika keluar dari scope
    
    let duration = start.elapsed();
    println!("Waktu eksekusi: {:?}", duration);
}